File: android/app/src/main/java/com/dailymemory/domain/usecase/memory/SaveMemoryUseCase.kt
package com.dailymemory.domain.usecase.memory

import com.dailymemory.domain.model.Memory
import com.dailymemory.domain.repository.MemoryRepository
import com.dailymemory.domain.repository.PersonRepository
import java.time.LocalDateTime
import javax.inject.Inject

/**
 * Use case for saving a new memory.
 * Handles the business logic of creating a memory and updating related persons.
 */
class SaveMemoryUseCase @Inject constructor(
    private val memoryRepository: MemoryRepository,
    private val personRepository: PersonRepository
) {
    suspend operator fun invoke(memory: Memory): Result<Memory> {
        return try {
            // Save the memory
            memoryRepository.insert(memory)

            // Update meeting count for related persons
            memory.personIds.forEach { personId ->
                personRepository.incrementMeetingCount(personId, memory.recordedAt)
            }

            Result.success(memory)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: ios/DailyMemory/Domain/UseCase/Memory/SaveMemoryUseCase.swift
import Foundation

/// Use case for saving a new memory.
/// Handles the business logic of creating a memory and updating related persons.
final class SaveMemoryUseCase {
    private let memoryRepository: MemoryRepository
    private let personRepository: PersonRepository
    private let semanticSearchUseCase: SemanticSearchUseCase?

    init(
        memoryRepository: MemoryRepository = MemoryRepositoryImpl(),
        personRepository: PersonRepository = PersonRepositoryImpl(),
        semanticSearchUseCase: SemanticSearchUseCase? = nil
    ) {
        self.memoryRepository = memoryRepository
        self.personRepository = personRepository
        self.semanticSearchUseCase = semanticSearchUseCase
    }

    func execute(_ memory: Memory) async throws -> Memory {
        // Save the memory
        try await memoryRepository.save(memory)

        // Update meeting count for related persons
        for personId in memory.personIds {
            try await personRepository.incrementMeetingCount(id: personId, date: memory.recordedAt)
        }

        // Index for semantic search (non-blocking)
        if let semanticSearch = semanticSearchUseCase {
            Task {
                await semanticSearch.indexMemory(memory)
            }
        }

        // Trigger cloud sync
        await SyncManager.shared.enqueueMemoryChange(id: memory.id, type: .create)

        return memory
    }
}
